enhancements/cluster-logging/logs-based-alerts.md

```diff
@@ -0,0 +1,300 @@
+---
+title: logs-based-alerts
+authors:
+  - "@periklis"
+reviewers:
+  - "@alanconway"
+  - "@xperimental"
+  - "@jcantrill"
+  - "@simonpasquier"
+  - "@fpetkovski"
+approvers:
+  - "@alanconway"
+  - "@simonpasquier"
+  - "@fpetkovski"
+api-approvers:
+  - "@alanconway"
+creation-date: "2022-05-16"
+last-updated: "2022-05-19"
+tracking-link:
+  - https://issues.redhat.com/browse/LOG-2510
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# Logs-based Alerts
+
+## Summary
+
+This document proposes a solution and a delivery plan for pushing logs-based alerts from OpenShift Logging Loki to Cluster-Monitoring AlertManager. It complements the existing [user-workload monitoring stack][uwm-docs] and [alerting stack][alerting-docs], enabling a full self-service experience for platform and workload logs-based monitoring.
```

Comment:
`plan for pushing logs-based alerts from OpenShift Logging Loki to Cluster-Monitoring AlertManager`

RulerConfig appears to handle this, routing to arbitrary AM instances.  But the other part of this enhancement is not about routing to AM, but presenting alerts via rules (AlertingRule CR).  Is this going to create a metric that can be scraped?  I ask because I want to make sure there isn't an assumption that AM is the only way to get log based alerts.  Specifically in the context of Managed OpenShift I have a few examples for why having AlertingRule create a metric akin to ALERTS is useful.

ROSA/OSD support log forwarding as an addon at this time.  There are no plans to SRE any full logging stack as a part of the platform.  Cluster monitoring is only used by SRE, customers cannot modify prometheus or AM configuration in openshift-monitoring.  Making sure customers can control and manage log based alerts includes flexibility in how the alerts are consumed.  And ARO only consumes ALERTS metrics from cluster prometheus.  ARO does not configure AM at all and never will due to use of internal MSFT systems as a first party offering in Azure.  If ARO were to ever consume logging metrics for the platform (not planned, just an example) we need the ALERTS metric, not AM routing.

Reply:
@jewzaam Do I understand correctly that specific SD use cases for alerting require that log-based queries to be transformed into metrics that get persisted into the cluster's Prometheus first?

Reply:
@jewzaam Scraping an alert metric for Prometheus is not going to work as per Loki executing LogQL expressions. However what we can do is to execute the LogQL query in Loki as a Recording rule and push the resulting metric to Prometheus via Remote Write. Basically Remote Write is the other half of the `RulerConfig` CRD and I aim to publish this as a separate document.

__ Note:__ The reason I started writing the Loki-to-AlertManager design document first is that it is from in-cluster logging perspective a low-hanging fruit. OTH Remote-Write into ClusterMonitoring/UserWorkloadMonitoring Prometheus requires a more robust design from reliability standpoint. I say this as per now Prometheus in either `openshift-monitoring` and `openshift-user-workload-monitoring` namespaces on any cluster is a closed black-box. Metrics can be exported via federation or Remote-Write to other services (in-/out-cluster) but you cannot write metrics into any of these Prometheus instances. For good reasons `openshift-monitoring` is closed for clients as it represents a first-line of defense operating the cluster itself, thus having Loki as writing client could overload it (e.g. high cardinality metrics from noisy/bad-crafted log queries).